### PR TITLE
fix(function): ST_Equals should return true for same type empty geometries

### LINKED ISSUE #26253: Mismatch between Presto Java and Presto C++ ST_EQUALS function
## Your Environment

* Presto version used:
* Storage (HDFS/S3/GCS..):
* Data source and connector used:
* Deployment (Cloud or On-prem):
* [Pastebin](https://pastebin.com/) link to the complete debug logs:

## Expected Behavior

Based on https://github.com/facebookincubator/velox/pull/15702#issuecomment-3811447646 We will implement the ST_Equals to be true for empty geometries regardless of their type.
presto docs: https://prestodb.io/docs/current/functions/geospatial.html#ST_Equals-Geometry-Geometry-boolean
velox docs: https://facebookincubator.github.io/velox/functions/presto/geospatial.html#ST_Equals

## Current Behavior

Presto side:

Image

Presto C++ side:

Image

Also looks like ANY empty even differing geometry types would return true on prestoc++ so even comparing `st_equals('polygon empty', 'linestring empty')` from the tables also gives true where presto gives false.

## Possible Solution

Assuming there is a mismatch with how the current libraries are dealing with empty cases. We can either check for empty cases and return true.

## Steps to Reproduce
1. shown in screenshot of java case how to create table and run the query.

## Context
Writing out end to end tests for ST_EQUALS comparing presto and prestoC++ and saw the mismatch.

### PR DESCRIPTION
**Pull request by jkhaliqi**
## Description

Presto ST_Equals function for empty geometries should return true regardless of geometry types. 

## Motivation and Context

Closes #26253 

## Test Plan

Added in tests.

## Contributor checklist

- [ ] Please make sure your submission complies with our [contributing guide](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md), in particular [code style](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style) and [commit standards](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#commit-standards).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.
- [ ] If adding new dependencies, verified they have an [OpenSSF Scorecard](https://securityscorecards.dev/#the-checks) score of 5.0 or higher (or obtained explicit TSC approval for lower scores).

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==
General Changes
* Update ST_Equals function for empty geometries to return true regardless of geometry types. 
* Add config properties for legacy ST_Equals behavior.
```

## Summary by Sourcery

Adjust ST_Equals geometry comparison semantics for empty geometries of the same or different types.

Bug Fixes:
- Ensure ST_Equals returns true for empty geometries of the same type and false for empty geometries of different types.

Tests:
- Add ST_Equals relation tests covering empty geometries of matching and differing types.

### CODE REVIEW

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs) 

Pull branch, local doc build, looks good. Thanks!

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 1]
```diff
@@ -730,3 +730,15 @@ The corresponding configuration property is :ref:`admin/properties:\`\`materiali
 
 Enable optimization to combine multiple :func:`!approx_distinct` function calls on expressions
 of the same type into a single aggregation using ``set_agg`` with array operations (``array_constructor``, ``array_transpose``).
+
+
+``legacy_st_equals``
```
Comment (czentgr) [inline presto-docs/src/main/sphinx/admin/properties-session.rst:1086]: This is currently listed under `View and Materialized View Properties`. But you need a new section if none exists for Geometries.

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 2]
```diff
@@ -1475,3 +1475,15 @@ Controls behavior when a materialized view is stale and no per-view staleness co
...
 The corresponding session property is :ref:`admin/properties-session:\`\`materialized_view_stale_read_behavior\`\``.
+
+``legacy-st-equals``
+^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+
+Enable legacy behavior for the ``ST_Equals`` geospatial function. This property
```
Comment (czentgr) [inline presto-docs/src/main/sphinx/admin/properties.rst]: What is the legacy behavior? And what is the new behavior (aka if not set)? We should explain it here or have a link with saying - check ST_equals behavior (link).

File: presto-main-base/src/test/java/com/facebook/presto/geospatial/TestGeoRelations.java [hunk 3]
```diff
@@ -155,6 +155,9 @@ public void testSTEquals()
         assertRelation("ST_Equals", "'MULTILINESTRING ((1 1, 5 1), (2 4, 4 4))'", "'MULTILINESTRING ((3 4, 6 4), (5 0, 5 4))'", false);
         assertRelation("ST_Equals", "'POLYGON ((1 1, 1 3, 3 3, 3 1, 1 1))'", "'POLYGON ((3 3, 3 1, 1 1, 1 3, 3 3))'", true);
         assertRelation("ST_Equals", "'MULTIPOLYGON (((1 1, 1 3, 3 3, 3 1, 1 1)), ((0 0, 0 2, 2 2, 2 0, 0 0)))'", "'POLYGON ((0 1, 3 1, 3 3, 0 3, 0 1))'", false);
+        assertRelation("ST_Equals", "'LINESTRING (0 0, 0 1)'", "'POINT EMPTY'", false);
```
Comment (czentgr) [inline presto-main-base/src/test/java/com/facebook/presto/geospatial/TestGeoRelations.java:161]: Can we add a a test for old behavior ? Aka in new behavior query passes but in old behavior it doesn't?

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 4]
```diff
@@ -932,3 +932,18 @@ When partition-aware execution is not applicable (for example: non-partitioned t
 columns in join conditions), the query falls back to standard grouped execution automatically.
 
 The corresponding configuration property is :ref:`admin/properties:\`\`partition-aware-grouped-execution-enabled\`\``.
+
+Geometry Properties
+-------------------
+
+``legacy_st_equals``Expand commentComment on line R852
```
Comment (czentgr) [inline presto-docs/src/main/sphinx/admin/properties-session.rst]: What's this? "Expand commentComment on line R852"

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 5]
```diff
@@ -1024,3 +1024,18 @@ When partition-aware execution is not applicable (for example: non-partitioned t
...
+
+``legacy_st_equals``
+^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+Enable legacy behavior for the ``ST_Equals`` geospatial function.
+See :func:`functions/geospatial:ST_Equals` for details on the behavior differences.
+This property will be removed in a future release.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties-session.rst]:
```suggestion
```
Documentation should not make promises of the future, they might not be kept.

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 6]
```diff
@@ -1786,3 +1786,19 @@ or HTTP/S. HTTPS are supported using the same internal communication HTTPS
...
+``legacy-st-equals``
+^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+
+Enable legacy behavior for the ``ST_Equals`` geospatial function.
+See :func:`functions/geospatial:ST_Equals` for details on the behavior differences.
+This property will be removed in a future release.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst]:
```suggestion
```

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 7]
```diff
@@ -1786,3 +1786,19 @@ or HTTP/S. HTTPS are supported using the same internal communication HTTPS
...
+
+``legacy-st-equals``
+^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+
+Enable legacy behavior for the ``ST_Equals`` geospatial function.
+See :func:`functions/geospatial:ST_Equals` for details on the behavior differences.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst]:
```suggestion
See ``ST_Equals`` in :ref:`functions/geospatial:Relationship Tests` for details on the behavior differences.
```
The link as given here doesn't work as a link, see this screenshot of a local doc build. 

Image

I can't manage to successfully format a link to https://prestodb.io/docs/current/functions/geospatial.html#ST_Equals-Geometry-Geometry-boolean directly, so this suggestion is a workaround that does compile successfully in a local build and will take the reader to where you want them to go.

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 8]
```diff
@@ -1024,3 +1024,18 @@ When partition-aware execution is not applicable (for example: non-partitioned t
...
+-------------------
+
+``legacy_st_equals``
+^^^^^^^^^^^^^^^^^^^^
+
+* **Type:** ``boolean``
+* **Default value:** ``false``
+
+Enable legacy behavior for the ``ST_Equals`` geospatial function.
+See :func:`functions/geospatial:ST_Equals` for details on the behavior differences.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties-session.rst]:
```suggestion
See ``ST_Equals`` in :ref:`functions/geospatial:Relationship Tests` for details on the behavior differences.
```
See the same comment in properties.rst for an explanation.

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 9]
```diff
@@ -142,7 +142,13 @@ Relationship Tests
 
 .. function:: ST_Equals(Geometry, Geometry) -> boolean
 
-    Returns ``true`` if the given geometries represent the same geometry.
+    Returns ``true`` if the given geometries represent the same geometry
+    according to ISO SQL/MM semantics. Also returns ``true`` if both geometries are empty,
+    regardless of their geometry types.
+    Note: A legacy implementation is available via the ``legacy_st_equals`` session property
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]:
```suggestion
    Note: A legacy implementation is available by using the ``legacy_st_equals`` session property
```
Avoid Latin abbreviations. See the [GitLab documentation style guide recommended word list entry for via](https://docs.gitlab.com/development/documentation/styleguide/word_list/#via) for discussion and alternatives.

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 10]
```diff
@@ -142,7 +142,13 @@ Relationship Tests
...
+    or ``legacy-st-equals`` configuration property. The legacy behavior returns ``false`` for
+    any two empty geometry comparisons, but otherwise behaves the same. This legacy behavior
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]:
```suggestion
    any two empty geometry comparisons, but otherwise behaves the same. 
```

File: presto-docs/src/main/sphinx/functions/geospatial.rst [hunk 11]
```diff
@@ -142,7 +142,13 @@ Relationship Tests
...
+    will be removed in a future release.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/functions/geospatial.rst]:
```suggestion
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the documentation!

File: presto-docs/src/main/sphinx/admin/properties.rst (hunk 6 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/admin/properties.rst]: This is still visible in a local doc build of the newly-pulled PR.

File: presto-docs/src/main/sphinx/admin/properties-session.rst (hunk 8 above)
Comment (steveburnett) [reply presto-docs/src/main/sphinx/admin/properties-session.rst]: This comment needs to be addressed.

File: presto-docs/src/main/sphinx/admin/properties.rst (hunk 6 above)
Comment (jkhaliqi) [reply presto-docs/src/main/sphinx/admin/properties.rst]: Updated, thank you!

File: presto-docs/src/main/sphinx/admin/properties-session.rst (hunk 8 above)
Comment (jkhaliqi) [reply presto-docs/src/main/sphinx/admin/properties-session.rst]: Updated, thank you!

**Review by steveburnett — DISMISSED** [review]
LGTM! (docs)

Please check that you've addressed and resolved @czentgr's comments - I think you have, but I didn't review those carefully to confirm them.

**Review by czentgr — DISMISSED** [review]
@jagill If you get a chance please review or maybe you have a suggestion for another reviewer?
@jaystarshot You might be interested in this as well.

File: presto-main-base/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java [hunk 12]
```diff
@@ -1089,11 +1090,16 @@ public static Boolean stDisjoint(@SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlTy
     @Description("Returns TRUE if the given geometries represent the same geometry")
     @ScalarFunction("ST_Equals")
     @SqlType(BOOLEAN)
-    public static Boolean stEquals(@SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlType(GEOMETRY_TYPE_NAME) Slice right)
+    public static Boolean stEquals(SqlFunctionProperties properties, @SqlType(GEOMETRY_TYPE_NAME) Slice left, @SqlType(GEOMETRY_TYPE_NAME) Slice right)
     {
         OGCGeometry leftGeometry = EsriGeometrySerde.deserialize(left);
         OGCGeometry rightGeometry = EsriGeometrySerde.deserialize(right);
         verifySameSpatialReference(leftGeometry, rightGeometry);
+        if (!properties.isLegacyStEquals()) {
```
Comment (tdcmeehan) [inline presto-main-base/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java]: This should not be in the hot path.  Instead, extract out this function into a standalone class, say, `LegacyStEquals`, and create a new one, `StEquals`.  Conditionally register either function based on `legacy_st_equals` in `BuiltInTypeAndFunctionNamespaceManager`.  For an example to refer to, see `use-new-nan-definition`.

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 13]
```diff
@@ -881,10 +881,10 @@ The value represents the max background fetch threads for refreshing metadata.
 * **Type:** ``boolean``
 * **Default value:** ``false``
 
-When enabled, the JDBC driver executes a validation query (``SELECT 1``) immediately 
-after establishing a connection to ensure it is working properly before returning it 
-to the application. This is useful for applications using connection pooling or 
-experiencing intermittent connectivity issues, as it allows early detection of failed 
+When enabled, the JDBC driver executes a validation query (``SELECT 1``) immediately
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/admin/properties-session.rst]: Please remove these whitespace changes

File: presto-docs/src/main/sphinx/admin/properties-session.rst [hunk 14]
```diff
@@ -900,8 +900,8 @@ Or set it by using connection properties::
 
 .. note::
 
-    Enabling connection validation adds a small overhead during connection establishment 
-    due to the execution of the validation query. This is typically negligible but should 
+    Enabling connection validation adds a small overhead during connection establishment
```
Comment (tdcmeehan) [inline presto-docs/src/main/sphinx/admin/properties-session.rst]: here too

**Review by tdcmeehan — COMMENTED** [review]
Just a couple of nits, otherwise LGTM

**Review by tdcmeehan — APPROVED** [review]
